mobify/platform-scaffold
# Force-load in preview, with PWA/non-PWA modes

#The V8 tag changed Preview behaviour. Previously, previewing force-loaded the loader (bypassed the user-agent check). Now, it doesn’t.

You *must* use the V8.1 tag when testing this PR. The V8.0 tag will load `loader.js` async under certain conditions (preview on Chrome 60, with a desktop UA),  and the loader will attempt to `document.write` a `<plaintext>` tag, which will fail because the DOM is already constructed. To use the new tag (if it hasn't been deployed to Merlins), use Charles with a Rewrite rule that replaces the existing tag (`` as a regex) with the V8 tag from the file `web/mobify-tag/v8.min.html` in this PR.

 **JIRA**: [WEBDATA-214](https://mobify.atlassian.net/browse/WEBDATA-214) which has been recreated as [TAG-1](https://mobify.atlassian.net/browse/TAG-1)

## Changes
- In `web/app/loader.js`:
  - Add `IS_PREVIEW` and `IS_PREVIEW_PWA_MODE` flags
  - Rename `IS_LOCAL_PREVIEW` to `IS_LOADED_LOCALLY` to better reflect the meaning of the flag
  - In `isSupportedPWABrowser`, return true if running in preview PWA mode, bypassing the UA check
  - Rename `isSupportedNonPWAMessagingBrowser` to `isSupportedNonPWABrowser` - nonPWA mode is not Messaging-specific
  - In `isSupportedNonPWABrowser`, if service workers are supported and running in preview nonPWA mode, return true, bypassing the UA check
- Update the V8 tag source files to V8.1

## Todos
- [ ] Update Merlins to the V8.1 tag *before merging and deploying this fix*

## How to test-drive this PR
- Run `npm run dev`
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fwww.merlinspotions.com%2F&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1) and verify that the loader is force-loaded in PWA mode, irrespective of the UA
- Add `non-pwa=1` to the hash part of the page URL (the preview parameters) and verify that the loader is force-loaded in non-PWA mode, irrespective of the UA
